sdks/java/io/hadoop-format/src/main/java/org/apache/beam/sdk/io/hadoop/format/HadoopFormatIO.java

```diff
@@ -0,0 +1,296 @@
+          TypeDescriptor.of(configuration.getClass(OUTPUTFORMAT_VALUE_CLASS, null));
+      Builder<K, V> builder =
+          toBuilder().setConfiguration(new SerializableConfiguration(configuration));
+      builder.setOutputFormatClass(outputFormatClass);
+      builder.setOutputFormatKeyClass(outputFormatKeyClass);
+      builder.setOutputFormatValueClass(outputFormatValueClass);
+
+      return builder.build();
+    }
+
+    /**
+     * Validates that the mandatory configuration properties such as OutputFormat class,
+     * OutputFormat key and value classes are provided in the Hadoop configuration.
+     */
+    private void validateConfiguration(Configuration configuration) {
+      checkArgument(configuration != null, "Configuration can not be null");
+      checkArgument(
+          configuration.get(OUTPUTFORMAT_CLASS) != null,
+          "Configuration must contain \"" + OUTPUTFORMAT_CLASS + "\"");
+      checkArgument(
+          configuration.get(OUTPUTFORMAT_KEY_CLASS) != null,
+          "Configuration must contain \"" + OUTPUTFORMAT_KEY_CLASS + "\"");
+      checkArgument(
+          configuration.get(OUTPUTFORMAT_VALUE_CLASS) != null,
+          "Configuration must contain \"" + OUTPUTFORMAT_VALUE_CLASS + "\"");
+    }
+
+    @Override
+    public void validate(PipelineOptions pipelineOptions) {}
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+      Configuration hadoopConfig = getConfiguration().get();
+      if (hadoopConfig != null) {
+        builder.addIfNotNull(
+            DisplayData.item(OUTPUTFORMAT_CLASS, hadoopConfig.get(OUTPUTFORMAT_CLASS))
+                .withLabel("OutputFormat Class"));
+        builder.addIfNotNull(
+            DisplayData.item(OUTPUTFORMAT_KEY_CLASS, hadoopConfig.get(OUTPUTFORMAT_KEY_CLASS))
+                .withLabel("OutputFormat Key Class"));
+        builder.addIfNotNull(
+            DisplayData.item(OUTPUTFORMAT_VALUE_CLASS, hadoopConfig.get(OUTPUTFORMAT_VALUE_CLASS))
+                .withLabel("OutputFormat Value Class"));
+      }
+    }
+
+    @Override
+    public PDone expand(PCollection<KV<K, V>> input) {
+      input.apply(ParDo.of(new WriteFn<>(this)));
+      return PDone.in(input.getPipeline());
+    }
+  }
+
+  private static class WriteFn<K, V> extends DoFn<KV<K, V>, Void> {
+    private final Write<K, V> spec;
+    private final SerializableConfiguration conf;
+    private transient RecordWriter<K, V> recordWriter;
+    private transient OutputCommitter outputCommitter;
+    private transient OutputFormat<?, ?> outputFormatObj;
+    private transient TaskAttemptContext taskAttemptContext;
+
+    WriteFn(Write<K, V> spec) {
+      this.spec = spec;
+      conf = spec.getConfiguration();
+    }
+
+    @Setup
+    public void setup() throws IOException {
+      if (recordWriter == null) {
+
+        taskAttemptContext = new TaskAttemptContextImpl(conf.get(), new TaskAttemptID());
```

Comment:
Using the default constructor here means the TaskID will default to 0 I think. 
I believe it is the case that temporary files can end up in `${mapreduce.output.fileoutputformat.outputdir}/_temporary/_${taskid}` and without a real `TaskID` is it possible there can be a file naming collision here?

See http://hadoop.apache.org/docs/r2.7.0/hadoop-mapreduce-client/hadoop-mapreduce-client-core/MapReduceTutorial.html#Task_Side-Effect_Files

Reply:
No it is not possible. We have this already implemented in upcoming PR. You either need to do shuffle or use external tool for synchronization (eg. hdfs).

Reply:
@dmvk Thanks. Which approach do you use in your PR?

Reply:
both, in streaming pipelines, you always need GBK, so there is no problem. In batch pipelines you can actually create random ids and optimistically create a lock file on hdfs, so you'll always get unique ids. If you don't have hdfs in place, you can fall back to GBK.

Reply:
@dmvk Ok, makes sense for me. 
Do you think you might update this PR with these changes or we keep it "as it is" for now and your upcoming PR will fix this?